### ISSUE TITLE
Target column with multiple classes (Multi Class Classification)

### ISSUE THREAD

**Issue by adidutt** [fairlearn/fairlearn#752]
Can we not implement bias detection in fairlearn if we have multiple classes in the target column like 0/1/2? What can be done to resolve this issue?

**Comment by hildeweerts** [comment]
Hi! Thank you for your question. 

It is possible to measure disparities for multi class classification problems, as long as the underlying metric supports multi class classification.

'Standard' group fairness metrics such as equalized odds (=differences in FPR and FNR) and demographic parity (=differences in selection rate) can be a bit tricky to compute for multi class classification scenarios, as "positives" and "negatives" are not directly defined. One way to deal with this is to view the problem as multiple binary classification problems and aggregate the results (see for example the different options for the 'average' parameter of [precision_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_score.html) documentation). We currently do **not** provide multi class false_positive_rate/FNR/selection_rate out of the box in Fairlearn.

However, scikit-learn does provide some performance metrics that can be applied to multi class classification problems, such as `accuracy_score` and `precision_score`. This may be relevant for you if you're trying to identify potential cases of quality of service harm (i.e., the model does not work equally well for all groups). You can use fairlearn's `MetricFrame` or `make_derived_metric` to explore disparities of these performance metrics across sensitive groups.  

For example, if we are interested to see if the model is less accurate for some groups compared to others, we can use scikit-learn's `accuracy_score` in combination with fairlearn's `MetricFrame` as follows:

```python
from sklearn.metrics import accuracy_score
from fairlearn.metrics import MetricFrame

# some random numbers
y_true = np.array([0,1,2,1,0,0,1,1])
y_pred = np.array([0,1,1,2,0,0,1,0])
sensitive_features = np.array([0,0,0,0,1,1,1,1])

# compute metric frame
mf = MetricFrame(metric=accuracy_score, 
                  y_true=y_true, y_pred=y_pred, 
                  sensitive_features=sensitive_features)
# print results
print(mf.by_group) # series with accuracy for each sensitive group
print(mf.difference()) # difference in accuracy between the two sensitive groups
```

Which will output:
```
sensitive_feature_0
0     0.5
1    0.75
Name: accuracy_score, dtype: object

0.25
```
Similarly, you could explore other performance metrics in a disaggregated way, such as recall, precision, AUC, etc. 

Does this answer your question? Please let me know if anything is still unclear.

**Comment by hildeweerts** [comment]
@fairlearn/fairlearn-maintainers do we want to add explicit mullticlass support for e.g., fpr and fnr? I am a bit worried that e.g., something like macro/weighted/etc. averaging would compress too much information in a single number. Alternatively we could write a section in the user guide to explain to people how they can compute a fairness metric for each class separately. wdyt?

**Comment by romanlutz** [comment]
We're facing the same in our chest x-ray classification use case and will just treat it as separate binary classification problems. I share your concern about condensing too much information into a number and I'm extremely (positively) surprised that we basically already support the accuracy score case!

A user guide section seems like the right choice. If others have ideas on how to handle this I'm more than willing to listen, though.

**Comment by hildeweerts** [comment]
>  I'm extremely (positively) surprised that we basically already support the accuracy score case!

The wonders of being scikit-learn compatible :D

**Comment by adrinjalali** [comment]
I'm happy to add fairness related metrics for multiclass cases, but metrics in those situations are tricky business. I'd be happy to include anything which is already at least half established in the community, rather than inventing our own ways of computing them.

For MetricFrame, as @hildeweerts mentions, as long as the underlying metric supports it, it's supported, I'd be happy with adding at least a section to the user guides explaining how users could use it for multi-class case.

**Comment by MiroDudik** [comment]
The idea of better inter-operability with confusion matrix has come up. One way how to enable it would be to have a bit of support for metrics that output pandas series (these would be automatically treated as multiple separate metrics).

Currently, we can do the following:
```python
import numpy as np
from sklearn.metrics import accuracy_score, confusion_matrix
from fairlearn.metrics import MetricFrame
import pandas as pd

# some random numbers
y_true = np.array([0,1,2,1,0,0,1,1])
y_pred = np.array([0,1,1,2,0,0,1,0])
sensitive_features = np.array([0,0,0,0,1,1,1,1])

# pandified confusion matrix with clarifying column annotations 
def confusion_matrix_pd(y_true, y_pred):
    matrix = confusion_matrix(y_true, y_pred, normalize='all')
    return pd.DataFrame(matrix).rename(
               index={0: "true=0", 1: "true=1", 2: "true=2"},
               columns={0: "pred=0", 1: "pred=1", 2: "pred=2"}).stack()

# construct metric frame
mf = MetricFrame(metric=confusion_matrix_pd, 
                  y_true=y_true, y_pred=y_pred, 
                  sensitive_features=sensitive_features)

# mf.by_group needs to be transformed to work properly
pd.DataFrame(mf.by_group.to_dict()).T.rename_axis(index='sensitive_feature_0')
```
Which outputs this:
```
                    true=0               true=1               true=2
                    pred=0 pred=1 pred=2 pred=0 pred=1 pred=2 pred=0 pred=1 pred=2
sensitive_feature_0
0                     0.25    0.0    0.0   0.00   0.25   0.25    0.0   0.25    0.0
1                     0.50    0.0    NaN   0.25   0.25    NaN    NaN    NaN    NaN
```
However, `mf.difference()` etc. will throw errors. It might be nice to basically enable API such that we would obtain the above without the transformation step (the last line above) and with `mf.difference()` etc. working columnwise.

**Comment by kumarnav03** [comment]
@hildeweerts  - If we are treating multi-class problem as multiple binary problem the could you help me understand how a new sample will be classified ? Are we going to take average or weighted of multiple binary result metric to classify?

**Comment by hildeweerts** [comment]
Hi @kumarnav03. Good question. What I meant was that you can treat it as multiple binary classification problems during the *evaluation* of the classifier. You can still use a multi class classifier to classify the samples. 

Does that make sense? Let me know if anything is still unclear!

**Comment by romanlutz** [comment]
Action item: add a short example like the one shown above by @hildeweerts to the user guide section on metrics with grouping: https://fairlearn.org/v0.6.2/user_guide/assessment.html#metrics-with-grouping

**Comment by SeanMcCarren** [comment]
Hi @romanlutz. I did this simple action item as my first issue, and I am running into some issues. I think I have to create a branch and then a pull request, but I don't have write access to fairlearn. Is this necessary or am I doing this wrong? :)

**Comment by romanlutz** [comment]
No worries! I admit it's something to get used to and I really should take some time to integrate the sprint instructions into our contributor guide. I'm glad you were able to figure it out! Looks good! I added a few comments but nothing major, so we should be able to merge this pretty soon.

**Comment by pradeepdev-1995** [comment]
@romanlutz  @hildeweerts  @SeanMcCarren   
Any updates on this thread? Shall we use fairlearn directly for mitigating bias in  multiclass classification problems?

**Comment by hildeweerts** [comment]
Hi, @pradeepdev-1995. Please note that this thread was mainly about measuring multiclass fairness, rather than mitigation. I'm closing this issue in favor of your issue #1201!
